### ISSUE TITLE
Lower Performance Compared to Java Client

### ISSUE THREAD

**Issue by HY1310** [Shopify/sarama#2103]
##### Versions

*Please specify real version numbers or git SHAs, not just "Latest" since that changes fairly regularly.*

| Sarama | Kafka | Go |
|--------|-------|----|
| Version 1.30.0 (2021-09-29) | 2.6 | 1.16 |

##### Configuration

What configuration values are you using for Sarama and Kafka?
``` go
        s.SaramaConfig = sarama.NewConfig()
	s.SaramaConfig.Version = sarama.V1_1_0_0
	s.SaramaConfig.Metadata.Timeout = time.Second * 30

	s.SaramaConfig.Producer.RequiredAcks = sarama.WaitForLocal
	

	s.SaramaConfig.Producer.Return.Successes = true
	s.SaramaConfig.Producer.Return.Errors = true

	s.SaramaConfig.Producer.Retry.Max = 0 // disable retry, use application level retry
	s.SaramaConfig.Producer.Flush.Bytes = 1024 * 1024
	s.SaramaConfig.Producer.Flush.Frequency = time.Millisecond
	s.SaramaConfig.Net.MaxOpenRequests = 5


	s.SaramaConfig.Producer.Flush.MaxMessages = 10
	s.SaramaConfig.Producer.Flush.Messages = 10
	s.SaramaConfig.Producer.MaxMessageBytes = 10 * 1024 * 1024
```

##### Problem Description
| client | qps |
|--------|-------|
| java | 1895 | 
| Sarma | 610 | 

test scenario,
message size - 1KB
network latency - 20ms

If the network latency is lower than 10ms Sarama will have similar qps as java client. But Sarama's performance decreases a lot when the network latency is high.

Looking into "async_producer.go" and "broker.go" we found that it's actually a sync invocation,
```go
func (b *Broker) Produce(request *ProduceRequest) (*ProduceResponse, error) {
	var (
		response *ProduceResponse
		err      error
	)

	if request.RequiredAcks == NoResponse {
		err = b.sendAndReceive(request, nil)
	} else {
		response = new(ProduceResponse)
		err = b.sendAndReceive(request, response)
	}

	if err != nil {
		return nil, err
	}

	return response, nil
}
```

This explains the qps decreasing. So is it possible to optimize the throughput of Sarama in high latency scenario? Thanks a lot.

**Comment by dnwe** [comment]
@HY1310 there was a proposed PR raised over Christmas that is intended to improve this particular situation. I haven’t yet had a chance to review it since returning from vacation, but if you could give it a try and see if it helps and report back that would be great!

see https://github.com/Shopify/sarama/pull/2094

**Comment by HY1310** [comment]
@dnwe I have tested the PR, the overall throughput of our application increased from about 500 qps to 2500 qps. It's awesome. Hope the PR can be merged soon.

**Comment by dnwe** [comment]
@HY1310 PR is merged and a new version [1.31.0](https://github.com/Shopify/sarama/releases/tag/v1.31.0) has been released, enojoy!

**Comment by HY1310** [comment]
@dnwe @slaunay thanks a lot:)
